katdal/caldata.py

```diff
@@ -0,0 +1,400 @@
+    return value is ``None``. Any keyword args are passed to
+    :func:`SimpleInterplate1D` or `ComplexInterpolate1D`.
+
+    Solutions that contain non-finite values are discarded.
+
+    Parameters
+    ----------
+    key : str
+        Name of the telescope state sensor
+
+    Returns
+    -------
+    interp : callable
+        Interpolation function which accepts timestamps and returns
+        interpolated data with shape (time, channel, pol, antenna). If the
+        solution is channel-independent, that axis will be present with
+        size 1.
+    """
+
+    timestamps, values = _get_cal_sensor(key, katdal_obj)
+
+    if values.ndim == 3:
+        # Insert a channel axis
+        values = values[:, np.newaxis, ...]
+
+    if values.ndim != 4:
+        raise CalibrationReadError('Calibration solutions has wrong number of dimensions')
+
+    # only use solutions with valid values, assuming matrix dimensions
+    # (ts, chan, pol, ant)
+    # - all values per antenna must be valid
+    # - all values per polarisation must be valid
+    # - some channels must be valid (you will interpolate over them later)
+    ts_mask = np.isfinite(values).all(axis=-1).all(axis=-1).any(axis=-1)
+    if (ts_mask.sum() / float(len(timestamps))) < 0.7:
+        raise CalibrationValueError('No finite solutions, not enough valid data')
+
+    # avoid extrapolation
+    if (timestamps[-1] < katdal_obj._timestamps[0]):
+        # All calibration solutions ahead of observation, use last one
+        values = values[[-1], ...]
+        timestamps = timestamps[[-1], ...]
+    elif (timestamps[0] > katdal_obj._timestamps[-1]):
+        # All calibration solutions after observation, use first one
+        values = values[[0], ...]
+        timestamps = timestamps[[0], ...]
+
+    values = values[ts_mask, ...]
+    timestamps = timestamps[ts_mask]
+
+    if values.shape[1] > 1:
+        # Only use channels selected in h5 file
+        values = values[:, katdal_obj.channels, ...]
+        for ts_idx in range(values.shape[0]):
+            # Interpolate across nans in channel axis.
+            for pol_idx in range(values.shape[2]):
+                for ant_idx in range(values.shape[3]):
+                    values[ts_idx, :, pol_idx, ant_idx] = interpolate_nans_1d(values[ts_idx, :, pol_idx, ant_idx])
+
+    # to interpolate you need >1 timestamps, else return only the single value
+    if timestamps.size > 1:
+        kind = kwargs.get('kind', 'linear')  # default if none given
+        if np.iscomplexobj(values) and kind not in ['zero', 'nearest']:
+            interp = ComplexInterpolate1D
+        else:
+            interp = SimpleInterpolate1D
+        return interp(timestamps, values)
+    return values
+
+
+class _cal_setup():
+    def __init__(self, katdal_obj):
+        def initcal(kind):
+            return {'interp': None, 'solns': None, 'kind': kind}
+        self._cal_solns = {'K': initcal('linear'),
```

Comment:
I think this should this be 'zero', because cal itself only does linear interpolation on G.